### PR TITLE
Better coverage

### PR DESCRIPTION
**Pull request by connectdotz**
fix a few test coverage issues (#429, #449, #488, #490, #499) and improve usability:

- the test coverage map only contains the tests in the latest run, which means in watch mode we lost the map generated during the earlier/initial test run. This leads to people had to pass `--watchAll` or restart the plugin in order to see coverage.
  - use `merge` to accumulate the maps instead of overriding.

- current coverage features display info from coverage-map but they do not help to generate the map. Users still need to pass the `--coverage` flag manually in either `jest.pathToJest`, jest config or package.json script to make it work. It is counter-intuitive and inefficient (why generate the map when toggle off the display, or toggle on display but do not generate the map?).
  - Giving `jest-editor-support` already has a `collectCoverage` option, we can easily wire it up with the existing features to provide the single-command, true-dynamic coverage experience out-of-the-box. 

- there is no indicator for coverage state, which can be confusing when the file is 100% covered. 
  - Added new "mode" icons in the status bar, currently shows "watch" and "coverage" state.

Screen Shot 2019-11-27 at 11 07 34 PM

### CODE REVIEW

**Comment by connectdotz** [comment]
@seanpoulter @stephtr @garyking when you guys get a chance, please take a look... hopefully, this will make our coverage support more intuitive going forward...

**Comment by connectdotz** [comment]
ping... anybody?

File: src/StatusBar.ts [hunk 1]
```diff
@@ -240,6 +246,22 @@ export class StatusBar {
...
+      switch (m) {
+        case 'coverage':
+          return '$(color-mode)'
+        case 'watch':
+          return '$(eye)'
+        default:
+          throw new Error(`unrecognized mode: ${m}`)
+      }
+    })
+    return modesStrings.length > 0 ? `${modesStrings.join(' ')}` : ''
```
Comment (stephtr) [inline src/StatusBar.ts]: Why not simply `return modesStrings.join(' ')`?

File: src/StatusBar.ts [hunk 2]
```diff
@@ -92,20 +95,20 @@ export class StatusBar {
...
       },
-      failed: (details?: string) => {
-        this.request(source, 'failed', details)
+      failed: (details?: string, modes?: Mode[]) => {
+        this.request(source, 'failed', details, modes)
       },
-      stopped: (details?: string) => {
-        this.request(source, 'stopped', details)
+      stopped: (details?: string, modes?: Mode[]) => {
+        this.request(source, 'stopped', details, modes)
```
Comment (stephtr) [inline src/StatusBar.ts]: This pattern feels a bit strange. What about instead adding a method to JestExt which calls `getStatusModes` and `request`?

Comment (connectdotz) [reply src/StatusBar.ts]: > What about instead adding a method to JestExt which calls getStatusModes and request?
right now mode and status are combined into a single string, if we make them separate calls, extra logic is needed to remember the missing info.

However, having said that, we could simplify the API by removing modes from all binding method except `initial` and injecting them in the wrapper within...

File: src/StatusBar.ts (hunk 1 above)
Comment (connectdotz) [reply src/StatusBar.ts]: indeed

File: src/StatusBar.ts (hunk 2 above)
Comment (connectdotz) [reply src/StatusBar.ts]: turns out the new API I was thinking about is even more cumbersome... maybe I should understand a bit more your concern so we can come up with a better approach...

**Comment by connectdotz** [comment]
> Maybe one can extend it in a future PR with a status bar button for switching the coverageOverlay.

that is a great idea, but you are right let's address it in another PR.

File: src/StatusBar.ts (hunk 2 above)
Comment (stephtr) [reply src/StatusBar.ts]: My approach would have been https://github.com/connectdotz/vscode-jest/pull/1.

**Pull request merged**